Update errors.md with info about Zones errors handling (#5612)
flutter/website: src/docs/testing/errors.md

The Flutter framework catches errors that occur during callbacks
triggered by the framework itself, including errors encountered

> during the build, layout, and paint phases. Errors which occur
> not within Flutter's callbacks cannot be caught by the framework,
> but you can handle them by setting up a Zone.

Reviewer: ```suggestion
during the build, layout, and paint phases. Errors that don't occur
within Flutter's callbacks can't be caught by the framework,
but you can handle them by setting up a [`Zone`][].
```

> All of errors caught by Flutter are routed to the

Reviewer: ```suggestion
All errors caught by Flutter are routed to the
```

in debug mode this shows an error message in red,
and in release mode this shows a gray background.

> When errors occur without a Flutter callback on the call
> stack they are handled by the Zone they occur in. By default
> Zones only print errors and do nothing else.

Reviewer: ```suggestion
When errors occur without a Flutter callback on the call stack,
they are handled by the `Zone` where they occur. By default,
a `Zone` only prints errors and does nothing else.
```

> they are handled by the <span class="x x-first x-last">`</span>Zone<span class="x x-first x-last">` where </span>they occur. By default<span class="x x-first x-last">,</span>

Author: That doesn't seem good :scream:

}
```

> ## Set up a Zone errors handling

Reviewer: ```suggestion
## Set up error handling for a Zone
```

Consider an `onPressed` callback that invokes an asynchronous function,

> such as `MethodChannel.invokeMethod` (or pretty much any plugin):

Reviewer: ```suggestion
such as `MethodChannel.invokeMethod` (or pretty much any plugin).
For example:
```

<!-- skip -->
```dart
OutlinedButton(
  child: Text('Click me!'),
  onPressed: () async {

>     final channel = const MethodChannel('crashy-custom-channel');

Reviewer: This made me LOL.  :D

    await channel.invokeMethod('blah');
  },
),
```

> If `invokeMethod` throws an error, it will not be forwarded to `FlutterError.onError`.

Reviewer: ```suggestion
If `invokeMethod` throws an error, it won't be forwarded to `FlutterError.onError`.
```

> Instead, it will be forwarded to the `Zone` where `runApp` was run.

Reviewer: ```suggestion
Instead, it's forwarded to the `Zone` where `runApp` was run.
```

To catch such an error, use [`runZonedGuarded`][].
<!-- skip -->
```dart
import 'dart:async';

void main() {
  runZonedGuarded(() async {
    WidgetsFlutterBinding.ensureInitialized();
    runApp(MyApp()));
  }, (Object error, StackTrace stack) {
    myBackend.sendError(error, stack);
  });
}
```

> Please note that `WidgetsFlutterBinding.ensureInitialized()` is
> called within `runZonedGuarded` - error handling wouldn't work if
> the function would be called from the outside.

Reviewer: ```suggestion
Note that `WidgetsFlutterBinding.ensureInitialized()` is
called within `runZonedGuarded`; error handling wouldn't work if
the function were called from the outside.
```

[`ErrorWidget.builder`]: {{site.api}}/flutter/widgets/ErrorWidget/builder.html
[`FlutterError.onError`]: {{site.api}}/flutter/foundation/FlutterError/onError.html
[`FlutterError.dumpErrorToConsole`]: {{site.api}}/flutter/foundation/FlutterError/dumpErrorToConsole.html
[`kReleaseMode`]:  {{site.api}}/flutter/foundation/kReleaseMode-constant.html
[`MaterialApp.builder`]: {{site.api}}/flutter/material/MaterialApp/builder.html
[reporting errors to a service]: /docs/cookbook/maintenance/error-reporting

> [`runZonedGuarded`]: {site.api}}/flutter/dart-async/runZonedGuarded.html

Reviewer: Add a [`Zone`] link to this section.

Also, the `runZonedGuarded` link is causing the build failure. It should be ``{{site.api}}``, not `{site.api}``.

}
```

## Set up error handling for a Zone

Consider an `onPressed` callback that invokes an asynchronous function,
such as `MethodChannel.invokeMethod` (or pretty much any plugin).
For example:
<!-- skip -->
```dart
OutlinedButton(
  child: Text('Click me!'),
  onPressed: () async {
    final channel = const MethodChannel('crashy-custom-channel');
    await channel.invokeMethod('blah');
  },
),
```
If `invokeMethod` throws an error, it won't be forwarded to `FlutterError.onError`.
Instead, it's forwarded to the `Zone` where `runApp` was run.

To catch such an error, use [`runZonedGuarded`][].
<!-- skip -->
```dart
import 'dart:async';

void main() {

>   runZonedGuarded(() async {

Reviewer 2: nit:
```suggestion
  runZonedGuarded(() {
```

    WidgetsFlutterBinding.ensureInitialized();

>     runApp(MyApp()));

Reviewer 2: ```suggestion
    runApp(MyApp());
```

>   }, (Object error, StackTrace stack) {

Reviewer 2: Is `WidgetsFlutterBinding.ensureInitialized()` really necessary here? If I'm not mistaken, it is also called from `runApp()`. I don't see why it needs to be called twice in a row in the same zone.

Author: It is indeed not necessary it this example.

But in many cases `WidgetsFlutterBinding.ensureInitialized()` needs to be called manually if a developer wants to initialize something before `runApp` (e.g. Crashlytics).

`WidgetsFlutterBinding.ensureInitialized()` is called manually in this example to emphasize that it needs to be called from inside of the `Zone`, otherwise errors handling won't work.

Reviewer 2: @blazern
Yes, but does it have to be inside `runZonedGuarded()` even if the initialization you mentioned is outside of it, or do you mean all initialization should be done inside `runZonedGuarded()` like below?

```dart
void main() {
  runZonedGuarded(() async {
    WidgetsFlutterBinding.ensureInitialized();
    await Firebase.initializeApp();
    runApp(MyApp());
  }, ...);
}
```

Author: @kaboc
I mean that if `WidgetsFlutterBinding.ensureInitialized()` is to be called, it must always be called inside `runZonedGuarded()`.

And if some initialization (like `await Firebase.initializeApp()`) must be called after `WidgetsFlutterBinding.ensureInitialized()`, then such initialization must also be called from `runZonedGuarded()`.

I thought that it can be easily understood from [the note below](https://github.com/flutter/website/pull/5612/files#diff-913f1e4f31d4c1700b06e4c3c4653f4470fe6c8f088b258fac6143a944767d3aR125)... Do you think the note should be reworded?

Reviewer 2: @blazern 
Now that I understand what you meant, the note looks easy to understand, but it was confusing at the first glance because I'd never seen pieces of code where initialization was done inside `runZonedGuarded()`.

How about adding some comment about initialization between `WidgetsFlutterBinding.ensureInitialized();` and `runApp(MyApp());` in the example code? I guess it'll help readers understand it more clearly.
